src/gui/elevation/qgselevationprofilecanvas.cpp

```diff
@@ -547,7 +638,94 @@ void QgsElevationProfileCanvas::generationFinished()
   zoomFull();
 }
 
-QgsProfilePoint QgsElevationProfileCanvas::canvasPointToPlotPoint( const QPointF &point ) const
+void QgsElevationProfileCanvas::onLayerProfileGenerationPropertyChanged()
+{
+  // TODO -- handle nicely when existing job is in progress
+  if ( !mCurrentJob || mCurrentJob->isActive() )
+    return;
+
+  QgsMapLayerElevationProperties *properties = qobject_cast< QgsMapLayerElevationProperties * >( sender() );
+  if ( !properties )
+    return;
+
+  if ( QgsMapLayer *layer = qobject_cast< QgsMapLayer * >( properties->parent() ) )
+  {
+    if ( QgsAbstractProfileSource *source = dynamic_cast< QgsAbstractProfileSource * >( layer ) )
+    {
+      if ( mCurrentJob->invalidateResults( source ) )
+        scheduleDeferredUpdate();
+    }
+  }
+}
+
+void QgsElevationProfileCanvas::onLayerProfileRendererPropertyChanged()
+{
+  // TODO -- handle nicely when existing job is in progress
+  if ( !mCurrentJob || mCurrentJob->isActive() )
+    return;
+
+  QgsMapLayerElevationProperties *properties = qobject_cast< QgsMapLayerElevationProperties * >( sender() );
+  if ( !properties )
+    return;
+
+  if ( QgsMapLayer *layer = qobject_cast< QgsMapLayer * >( properties->parent() ) )
+  {
+    if ( QgsAbstractProfileSource *source = dynamic_cast< QgsAbstractProfileSource * >( layer ) )
+    {
+      mCurrentJob->replaceSource( source );
+    }
+    if ( mPlotItem->redrawResults( layer->id() ) )
+      scheduleDeferredRedraw();
+  }
+}
+
+void QgsElevationProfileCanvas::updateResultsForLayer()
+{
+  if ( QgsMapLayer *layer = qobject_cast< QgsMapLayer * >( sender() ) )
+  {
+    if ( QgsAbstractProfileSource *source = dynamic_cast< QgsAbstractProfileSource * >( layer ) )
+    {
+      if ( mCurrentJob->invalidateResults( source ) )
+        scheduleDeferredUpdate();
+    }
+  }
+}
+
+void QgsElevationProfileCanvas::scheduleDeferredUpdate()
+{
+  if ( !mDeferredUpdateScheduled )
+  {
+    mDeferredUpdateTimer->start( 1 );
+    mDeferredUpdateScheduled = true;
+  }
+}
+
+void QgsElevationProfileCanvas::scheduleDeferredRedraw()
+{
+  if ( !mDeferredRedrawScheduled )
+  {
+    mDeferredRedrawTimer->start( 1 );
+    mDeferredRedrawScheduled = true;
+  }
+}
+
+void QgsElevationProfileCanvas::startDeferredUpdate()
+{
+  if ( mCurrentJob && !mCurrentJob->isActive() )
+  {
+    mCurrentJob->updateInvalidatedResults();
+  }
+
+  mDeferredUpdateScheduled = false;
+}
+
+void QgsElevationProfileCanvas::startDeferredRedraw()
+{
+  mPlotItem->update();
+  mDeferredRedrawScheduled = false;
+}
```

Comment:
I have been getting a bit confused what's the difference between an "update" and a "redraw" - maybe worth commenting somewhere which does what...

Reply:
Ok, I've renamed all the references to "update" meaning "rebuild the profile for a layer" to "regenerate", so that we don't get the naming clash with Qt's "update=repaint" association.